### PR TITLE
[MM-15477][MM-27540][MM-27541]: fix search inconsistencies using `in:@` / `in:`

### PR DESCRIPTION
**Pull request by ashishbhate**
#### Summary

- Fix ordering of `in:` suggestions: DMs, GMs, Private and then Public channels
- Return only DMs and GMs for `in:@` search, unless a channel name/display name also starts with `@`, in which case it will also be included.

#### Ticket Link

- https://mattermost.atlassian.net/browse/MM-15477
- https://mattermost.atlassian.net/browse/MM-27540
- https://mattermost.atlassian.net/browse/MM-27541

#### Checklist

- [x] Added or updated unit tests (required for all new features)

#### Device Information
This PR was tested on:  
- There is some issue with my local setup. I'm not able to run apps through the adb bridge or build an apk. I'm still trying to figure this out.
For now I've tested in an Android Emulator - Pixel 3 XL running Android 11

#### Screenshots

Screenshot_1597925024

### CODE REVIEW

File: app/selectors/autocomplete.js [hunk 1]
```diff
@@ -258,14 +258,18 @@ export const filterDirectAndGroupMessages = createSelector(
         if (matchTerm === null) {
             return null;
         }
+        let mt = matchTerm;
+        if (matchTerm.startsWith('@')) {
+            mt = matchTerm.replace(/^@/, '');
```
Comment (enahum) [inline app/selectors/autocomplete.js:263]: you know that `matchTerm` starts with and `@` sign, so using a substr is faster than replacing with a regex so you could do

```suggestion
            mt = matchTerm.substr(1);
```

**Comment by enahum** [comment]
I guess that is a server problem returning the results and not this PR, I am right ?

**Comment by josephbaylon** [comment]
@enahum Web is returning the results properly. I based the comparison on the web.

**Comment by enahum** [comment]
Then I wonder if it is an issue with the filters applied in the PR.

Thanks @josephbaylon for clarifying ;)

**Comment by ashishbhate** [comment]
@josephbaylon thanks for testing.

>It looks like search does not return results when term is in the middle.

I just tested with the master branch. It appears that in the master branch as well the search does not return results when the term is in the middle. Can you confirm?

**Comment by josephbaylon** [comment]
@ashishbhate At least on v1.35 build 319, search does not return results when channel/gm has prefix. But wasn't sure if your PR should be fixing that behavior as well. If not I can approve, we'll just have to indicate in the PR.

**Comment by ashishbhate** [comment]
@josephbaylon the intent is only to fix the ordering. I think the effort to improve search overall is being tracked in https://mattermost.atlassian.net/browse/MM-26724

This difference in webapp and mobile search results is definitely something we should track as well. I'll poke around a bit more to figure out why its happening exactly. I'll also bring it up during the core features team meeting and maybe the mobile team meeting.

**Comment by josephbaylon** [comment]
@ashishbhate Thanks! I'll wait on your comment on what's decided in the meetings before I approve.

**Comment by ashishbhate** [comment]
@josephbaylon as discussed in the meetings - I've opened a asana ticket to track the discrepancies in the webapp and mobile auto-suggest results: https://mattermost.atlassian.net/browse/MM-28241

**Comment by josephbaylon** [comment]
Hi @ashishbhate are you going to address the fix for the ticket partly in this PR or in a totally separate PR(s)?

**Comment by ashishbhate** [comment]
In totally separate PRs. The fix will take some time and might need a lot of changes.

**Pull request merged**